libbeat/statestore/backend/memlog/diskstore.go

```diff
@@ -165,5 +264,410 @@ func (s *diskstore) LogOperation(op op) error {
 // NOTE: due to limitation on some Operating system or file systems, the active
 //       marker is not a symlink, but an actual file.
 func (s *diskstore) WriteCheckpoint(state map[string]entry) error {
-	panic("TODO: implement me")
+	tmpPath, err := s.checkpointTmpFile(filepath.Join(s.home, "checkpoint"), state)
+	if err != nil {
+		return err
+	}
+
+	fileTxID := s.txid + 1
+	fileName := fmt.Sprintf("%v.json", fileTxID)
+	checkpointPath := filepath.Join(s.home, fileName)
+
+	if err := os.Rename(tmpPath, checkpointPath); err != nil {
+		return err
```

Comment:
If there's an error here, should we delete the temporary file?

Related, it looks like `checkpointTmpFile` always uses the file name `"checkpoint.new"`. This means that if a checkpoint is incomplete (whether because of an explicit error or just incomplete cleanup), it might obstruct the creation of new ones (this will usually not be the case since you use `O_TRUNC` etc but it could happen e.g. if the beat is restarted by a different user than previously, or if file permissions are otherwise changed). It would be a bit more robust to include the txid in the temporary file as well, so there will ~never be a conflict.

Reply:
Restarting as another user that doesn't have access should trigger an error early when opening the store.

Changing permissions life can indeed be a risk, even for this file. I opted to not include the txid in the temporary file, as we continuously will need to try to cleanup if there was some error in the meantime. Messing with permissions on existing files or the directory can lead to other problems as well, so I didn't really account for that. In order to prevent a potential attack on the temporary file I wouldn't use the `txid` for the temporary file, but really create a full random name.

I will add logic to delete an existing `checkpoint.new` on init time (this is in `openStore` which is not shown here), so we can fail with permission errors on Beats start. Plus I will add a os.Remove that tries to delete the `checkpoint.new` file if it still exists after the renaming. Any concerns with that?